hera_cal/abscal.py

```diff
@@ -1291,6 +1295,280 @@ def global_phase_slope_logcal(model, data, antpos, reds=None, solver='linfit', w
         return gains
 
 
+@jax.jit
+def _eval_Z(x, b_vecs, normalized_data_model_ratio, weights):
+    """
+    Computes a metric of how well the model and data can be made to match after abscal.
+
+    Parameters
+    ----------
+    x : np.ndarray or jnp.ndarray
+        Solution vector of the tip-tilt phase slopes. Shape (n_dims,).
+    b_vecs : np.ndarray or jnp.ndarray
+        Array of baseline vectors. Shape (n_bls, n_dims).
+    normalized_data_model_ratio : np.ndarray or jnp.ndarray
+        Array of normalized data/model ratios. Shape (n_bls,).
+    weights : np.ndarray or jnp.ndarray
+        Array of weights. Shape (n_bls,).
+    
+    Returns
+    -------
+    Z : float
+        Metric of how well the model and data can be made to match after abscal.
+    """
+    phase = jnp.sum(x * b_vecs, axis=1)
+    phasor = jnp.cos(phase) + 1j * jnp.sin(phase)
+    Z = jnp.mean(phasor * normalized_data_model_ratio * weights)
+    return Z
+
+@jax.jit
+def _grad_and_hess_real(x, b_vec, w_expi_psi):
+    """
+    Evaluates the gradient and hessian for a given set of input parameters, x.
+
+    Parameters:
+    -----------
+    x : np.ndarray or jnp.ndarray
+        Solution vector of the tip-tilt phase slopes. Shape (n_dims,).
+    b_vec : np.ndarray or jnp.ndarray
+        Array of baseline vectors. Shape (n_bls, n_dims).
+    w_expi_psi : np.ndarray or jnp.ndarray
+        Array of weighted normalized data/model ratios. Shape (n_bls,).
+
+    Returns:
+    --------
+    grad : np.ndarray or jnp.ndarray
+        Gradient of the objective function. Shape (n_dims,).
+    hess : np.ndarray or jnp.ndarray
+        Hessian of the objective function. Shape (n_dims, n_dims).
+    """
+    x_dot_b = jnp.dot(b_vec, x)
```

Comment:
Might be good to add comments (or more to the docstring) explaining the analytics here and why these sines and cosines are the right thing to do.

Reply:
Will do

Reply:
avoiding evaluating complex exponentials/creating complex arrays with a little trig